### PR TITLE
Commit made atomic

### PR DESCRIPTION
**Pull request by Arghya115**

### CODE REVIEW

File: src/Lachain.Storage/State/StateManager.cs [hunk 1]
```diff
@@ -21,8 +20,9 @@ public class StateManager : IStateManager
         private readonly ISnapshotManager<IValidatorSnapshot> _validatorManager;
 
         private readonly Mutex _globalMutex = new Mutex(false);
+        public IRocksDbContext _dbContext;
```
Comment (HMAshiqulIslam) [inline src/Lachain.Storage/State/StateManager.cs:23]: should it be public ?

File: src/Lachain.Storage/StorageState.cs [hunk 2]
```diff
@@ -104,12 +104,12 @@ public ulong TryDelete(byte[] key, out byte[]? value)
...
 
         [MethodImpl(MethodImplOptions.Synchronized)]
-        public ulong Commit()
+        public ulong Commit(RocksDbAtomicWrite batch)
         {
-            using var tx = _repositoryManager.CreateTransaction();
-            _trieMap.Checkpoint(CurrentVersion, tx);
-            _repositoryManager.SetState(CurrentVersion, tx);
-            tx.Commit();
+        //    using var tx = _repositoryManager.CreateTransaction();
```
Comment (HMAshiqulIslam) [inline src/Lachain.Storage/StorageState.cs:109]: what does this comments mean ?

File: src/Lachain.Storage/State/StateManager.cs [hunk 3]
```diff
@@ -143,13 +145,20 @@ public void Rollback()
...
+            _contractManager.Commit(batch);
+            _storageManager.Commit(batch);
+            _transactionManager.Commit(batch);
+            _blockManager.Commit(batch);
+            _eventManager.Commit(batch);
+            _validatorManager.Commit(batch);
+            batch.Commit();
+        }
+
+        public void Commit(RocksDbAtomicWrite batch)
```
Comment (HMAshiqulIslam) [inline src/Lachain.Storage/State/StateManager.cs:159]: why do we need this ?

**Review by HMAshiqulIslam — CHANGES_REQUESTED** [review]

File: src/Lachain.Storage/State/StateManager.cs (hunk 3 above)
Comment (Arghya115) [reply src/Lachain.Storage/State/StateManager.cs:159]: StateManager inherits from IStateManager and IStateManager inherits from ISnapshotManager<IBlockchainSnapshot>, now SnapshotManager also inherits from ISnapshotManager. I needed to put Commit(batch) in the ISnapshotManager and as a result I also had to put it in StateManager. This looks really stupid but I can't think of any other way without massive change in the architecture.

File: src/Lachain.Storage/StorageState.cs (hunk 2 above)
Comment (Arghya115) [reply src/Lachain.Storage/StorageState.cs:109]: Removing this now.

File: src/Lachain.Storage/State/StateManager.cs (hunk 1 above)
Comment (Arghya115) [reply src/Lachain.Storage/State/StateManager.cs:23]: Not sure if there is any problem with public, but keeping it private is better idea. Changing it now.

File: src/Lachain.Storage/State/StateManager.cs (hunk 3 above)
Comment (HMAshiqulIslam) [reply src/Lachain.Storage/State/StateManager.cs:159]: maybe we can add some comments or throw exception inside this method to make sure none uses it accidentally.

**Review by sgladkov — APPROVED** [review]

**Pull request merged**